File: ProblemSet_13xx/Problem_1339/solution.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def maxProduct(self, root: Optional[TreeNode]) -> int:
        my_sum = 0
        
        def traverse(node: TreeNode):
            nonlocal my_sum
            my_sum += node.val
            
            if (node.left):
                traverse(node.left)
                
            if (node.right):
                traverse(node.right)
        
        traverse(root)
        
        result = 0
        def traverse_cut(node: TreeNode) -> int:
            nonlocal result
            left = 0
            if (node.left != None):
                left = traverse_cut(node.left)
                remaining = my_sum - left
                result = max(result, remaining * left)
            
            right = 0
            if (node.right != None):
                right = traverse_cut(node.right)
                remaining = my_sum - right
                result = max(result, remaining * right)
            
            return left + right + node.val
        
        traverse_cut(root)
        return result % (10**9 + 7)

File: ProblemSet_13xx/Problem_1339/solution.cs
using System.Xml;

Solution s = new();
TreeNode t = new TreeNode(1,
    new TreeNode(2,
        new TreeNode(4),
        new TreeNode(5)
    ),
    new TreeNode(3,
        new TreeNode(6)
    )
);
Console.WriteLine(s.MaxProduct(t));

t = new TreeNode(1,
    new TreeNode(2,
        new TreeNode(3),
        new TreeNode(4,
            new TreeNode(5),
            new TreeNode(6)
        )
    )
);
Console.WriteLine(s.MaxProduct(t));


public class TreeNode {
    public int val;
    public TreeNode left;
    public TreeNode right;
    public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}
public class Solution {
    public int sum;
    public long max;
    public int MaxProduct(TreeNode root)
    {
        max = 0;
        sum = SumTree(root);
        SumWithProd(root);
        return (int)(max % 1000000007);
    }
    public int SumTree(TreeNode root)
    {
        if (root == null) return 0;
        return root.val + SumTree(root.left) + SumTree(root.right);
    }
    public long SumWithProd(TreeNode root)
    {
        if (root == null) return 0;
        long left = SumWithProd(root.left);
        long right = SumWithProd(root.right);

        max = Math.Max(max, left * (sum - left));
        max = Math.Max(max, right * (sum - right));
        return root.val + left + right;
    }
}